File: apps/remote_browser/viewer.py
"""Uzak tarayıcının tabletteki görünümü (Streamlit components v2; JS: `viewer.js`).

İki bileşen: ortada ekran + sağda yazı paneli (dokunma → tıklama, sürükleme / tekerlek → kaydırma, yazı → klavye) ve
kenar çubuğunda gezinme, adres, sekmeler, indirilenler. Olaylar `input` durumuyla Python'a döner, `apply_events`
tarayıcıya uygular.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import streamlit as st

from .service import KEYS, VIEWPORT, RemoteBrowser

ICON = '<svg viewBox="0 0 24 24" aria-hidden="true">{}</svg>'
ICONS = {
    "back": ICON.format('<path d="M15 18l-6-6 6-6"/>'),
    "forward": ICON.format('<path d="M9 18l6-6-6-6"/>'),
    "reload": ICON.format('<path d="M20 11a8 8 0 1 0-2.3 5.7"/><path d="M20 4v7h-7"/>'),
    "home": ICON.format('<path d="M3 11l9-8 9 8"/><path d="M5 10v10h5v-6h4v6h5V10"/>'),
}

# Ortada ekran (4:3), sağda yazı paneli; tablet dik tutulursa panel ekranın altına iner.
SCREEN_HTML = """
<div class="rb">
  <div class="screen-wrap"><img class="screen" alt="Tarayıcı" tabindex="0" draggable="false"><span class="wait"></span>
    <div class="login"></div></div>
  <aside class="side">
    <div class="h">Yazı</div>
    <form class="typebar">
      <input class="text" type="text" autocomplete="off" placeholder="Önce ekranda kutuya dokun">
      <button type="submit" class="primary">Yaz</button>
    </form>
    <div class="keys">
      <button type="button" data-k="Enter" title="Enter">↵</button><button type="button" data-k="Backspace" title="Sil">⌫</button>
      <button type="button" data-k="Tab" title="Sonraki kutu">⇥</button>
    </div>
    <button type="button" class="pw" data-t="password" title="Bu sitenin kayıtlı kullanıcı adı ve şifresini giriş kutularına yazar">🔑 Girişi doldur</button>
  </aside>
</div>
"""

PANEL_HTML = f"""
<div class="pn">
  <div class="nav">
    <button type="button" data-t="back" title="Geri">{ICONS["back"]}</button>
    <button type="button" data-t="forward" title="İleri">{ICONS["forward"]}</button>
    <button type="button" data-t="reload" title="Yenile">{ICONS["reload"]}</button>
    <button type="button" data-t="home" title="DHA haberleri">{ICONS["home"]}</button>
  </div>
  <form class="go"><input class="url" type="text" inputmode="url" autocomplete="off" spellcheck="false"
    placeholder="Adres veya arama"><button type="submit" class="primary">Git</button></form>
  <div class="h">Sekmeler</div><div class="tabs"></div>
  <div class="h">İndirilenler</div><div class="downloads"></div>
</div>
"""

BASE_CSS = """
:host, .rb, .pn { --navy: #123249; --sky: #BEE1E8; --lime: #D0E491; --soft: #F3F7FA; --line: #DCE5EC; --ink: #1B2B3A;
  font-family: "Source Sans Pro", system-ui, sans-serif; color: var(--ink); }
button { font: inherit; border: 1px solid var(--line); background: #fff; color: var(--navy); border-radius: 10px;
  min-height: 40px; padding: 0 12px; font-size: 15px; font-weight: 600; cursor: pointer; transition: background .12s, transform .06s; }
button:active { background: #E7F1FB; transform: scale(.97); }
button.primary { background: var(--navy); border-color: var(--navy); color: #fff; }
button.primary:active { background: #0B2233; }
input { font: inherit; min-height: 40px; border: 1px solid var(--line); border-radius: 10px; padding: 0 10px; font-size: 16px;
  min-width: 0; background: #fff; color: var(--ink); outline: none; }
input:focus { border-color: var(--navy); box-shadow: 0 0 0 3px rgba(190, 225, 232, .7); }
.h { font-size: 12px; font-weight: 700; letter-spacing: .06em; text-transform: uppercase; color: #6B7C8C; margin: 4px 2px -2px; }
"""

SCREEN_CSS = BASE_CSS + """
.rb { display: grid; grid-template-columns: minmax(0, 1fr) 156px; gap: 12px; align-items: start; }
.screen-wrap { position: relative; width: 100%; max-width: calc((100vh - 36px) * 4 / 3); justify-self: center;
  border-radius: 12px; overflow: hidden; box-shadow: 0 1px 2px rgba(18, 50, 73, .08), 0 6px 20px rgba(18, 50, 73, .10);
  background: var(--soft); }
.screen { display: block; width: 100%; aspect-ratio: var(--ar, 4 / 3); touch-action: none; user-select: none;
  -webkit-user-select: none; outline: none; will-change: transform; }
.screen-wrap:has(.screen:focus-visible) { box-shadow: 0 0 0 3px var(--sky); }
.wait { position: absolute; right: 10px; top: 10px; width: 16px; height: 16px; border-radius: 50%;
  border: 2px solid var(--sky); border-top-color: var(--navy); animation: spin .8s linear infinite; display: none; }
.wait.on { display: block; }
@keyframes spin { to { transform: rotate(360deg); } }
.ripple { position: absolute; width: 28px; height: 28px; margin: -14px 0 0 -14px; border-radius: 50%;
  background: rgba(18, 50, 73, .28); pointer-events: none; animation: rip .45s ease-out forwards; }
@keyframes rip { to { transform: scale(1.8); opacity: 0; } }
.side { display: flex; flex-direction: column; gap: 8px; position: sticky; top: 8px; }
.typebar { display: flex; flex-direction: column; gap: 8px; }
.keys { display: grid; grid-template-columns: repeat(3, 1fr); gap: 6px; }
.keys button { padding: 0; font-size: 18px; }
/* Giriş çubuğu görüntünün üstüne biner (alt kenar): belirince görüntü kaymaz, dokunuşlar yanlış yere gitmez. */
.login:empty { display: none; }
.login { position: absolute; left: 10px; right: 10px; bottom: 10px; display: flex; gap: 8px; align-items: center;
  flex-wrap: wrap; padding: 10px 12px; border-radius: 12px; background: #FDF1DC; font-size: 15px;
  box-shadow: 0 4px 16px rgba(18, 50, 73, .25); }
.login.filled { background: #E6F4EC; }
.login span { flex: 1; min-width: 200px; }
.login button { min-height: 36px; }
@media (orientation: portrait) {
  .rb { grid-template-columns: 1fr; }
  .screen-wrap { max-width: 100%; }
  .side { position: static; display: grid; grid-template-columns: 1fr auto auto; align-items: end; }
  .side .h { display: none; }
  .typebar { flex-direction: row; }
}
"""

PANEL_CSS = BASE_CSS + """
.pn { display: flex; flex-direction: column; gap: 8px; }
.nav { display: grid; grid-template-columns: repeat(4, 1fr); gap: 6px; }
.nav button { padding: 0; display: grid; place-items: center; }
.nav svg { width: 20px; height: 20px; fill: none; stroke: currentColor; stroke-width: 2.4; stroke-linecap: round; stroke-linejoin: round; }
.go { display: flex; gap: 6px; }
.go input { flex: 1; width: 0; }
.go button { padding: 0 10px; }
.tabs { display: flex; flex-direction: column; gap: 4px; }
.tab { display: flex; align-items: stretch; border: 1px solid var(--line); border-radius: 10px; background: #fff; overflow: hidden; }
.tab.on { border-color: var(--navy); background: #EAF3F8; box-shadow: inset 3px 0 0 var(--navy); }
.tab button { border: 0; border-radius: 0; background: transparent; min-height: 38px; }
.tab .pick { flex: 1; min-width: 0; text-align: left; font-weight: 500; font-size: 14px; white-space: nowrap; overflow: hidden;
  text-overflow: ellipsis; padding: 0 10px; }
.tab.on .pick { font-weight: 700; }
.tab .x { width: 36px; padding: 0; color: #6B7C8C; font-size: 14px; }
.downloads { display: flex; flex-direction: column; gap: 6px; font-size: 14px; }
.dl { padding: 8px 10px; border-radius: 10px; background: var(--soft); border: 1px solid var(--line); display: flex;
  flex-direction: column; gap: 4px; }
.dl .name { white-space: nowrap; overflow: hidden; text-overflow: ellipsis; }
.dl .info { font-size: 12px; color: #6B7C8C; }
.dl.bitti { background: #EEF7F1; border-color: #CBE7D5; } .dl.hata { background: #FDECEA; border-color: #F4C7C1; }
.dl button { min-height: 34px; font-size: 13px; padding: 0 8px; }
.empty { font-size: 13px; color: #6B7C8C; padding: 2px; }
"""

JS = Path(__file__).with_name("viewer.js").read_text(encoding="utf-8")
_MOUNTS: dict[str, Any] = {}
MAX_EVENTS = 60


def _component(name: str, **parts: str):
    """Bileşen bu Streamlit çalışma ortamında kayıtlı değilse kaydeder (testlerde her uygulama yeni ortamdır)."""
    from streamlit.components.v2 import get_bidi_component_manager

    if name not in _MOUNTS or get_bidi_component_manager().get(name) is None:
        _MOUNTS[name] = st.components.v2.component(name, **parts)
    return _MOUNTS[name]


def _number(value: Any, low: float, high: float) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return min(high, max(low, number)) if number == number else None


def apply_events(browser: RemoteBrowser, events: Any, password: str | None, home: str) -> int:
    """Tabletten gelen olayları doğrulayıp tarayıcıya uygular; uygulananların sayısını döndürür."""
    applied = 0
    width, height = VIEWPORT
    for event in (events if isinstance(events, list) else [])[:MAX_EVENTS]:
        if not isinstance(event, dict):
            continue
        kind, value = event.get("t"), event.get("v")
        if kind in ("click", "wheel") and isinstance(value, list) and len(value) >= 2:
            x, y = _number(value[0], 0, width - 1), _number(value[1], 0, height - 1)
            if x is None or y is None:
                continue
            if kind == "click":
                browser.run("click", (x, y))
            else:
                delta = _number(value[2] if len(value) > 2 else 0, -6000, 6000)
                if not delta:
                    continue
                browser.run("wheel", (x, y, delta))
        elif kind == "goto" and isinstance(value, str) and value.strip():
            browser.run("goto", value[:2000])
        elif kind == "home":
            browser.run("goto", home)
        elif kind == "type" and isinstance(value, str) and value:
            browser.run("type", value[:2000])
        elif kind == "key" and value in KEYS:
            browser.run("key", value)
        elif kind in ("back", "forward", "reload"):
            browser.run(kind)
        elif kind in ("tab", "close_tab"):
            index = _number(value, 0, 50) if value is not None else None
            if kind == "tab" and index is None:
                continue
            browser.run(kind, None if index is None else int(index))
        elif kind == "password":
            # Kayıtlı giriş varsa kutular doldurulur; yoksa (eski ayar) DHA_SIFRE seçili kutuya yazılır.
            if not browser.fill_login() and password:
                browser.run("type", password)
        elif kind in ("save_login", "dismiss_login"):
            browser.answer_login_offer(kind == "save_login")
        else:
            continue
        applied += 1
    return applied


def browser_view(data: dict[str, Any], key: str) -> dict[str, Any] | None:
    """Ekran + yazı paneli (ana alan)."""
    mount = _component("axion_uzak_tarayici_ekran", html=SCREEN_HTML, css=SCREEN_CSS, js=JS)
    result = mount(data={**data, "viewport": list(VIEWPORT)}, key=key, default={"input": None}, on_input_change=lambda: None)
    return result.input


def panel_view(data: dict[str, Any], key: str) -> dict[str, Any] | None:
    """Gezinme, adres, sekmeler, indirilenler (kenar çubuğu)."""
    mount = _component("axion_uzak_tarayici_panel", html=PANEL_HTML, css=PANEL_CSS, js=JS)
    result = mount(data=data, key=key, default={"input": None}, on_input_change=lambda: None)
    return result.input

File: apps/remote_browser/viewer.js
// Uzak tarayıcının tabletteki görünümü. Python ekran görüntüsünü (JPEG) yollar; bu dosya dokunuşları tarayıcının
// koordinatlarına çevirip olay olarak geri yollar (`input`). Dokun = tıkla, sürükle / tekerlek = kaydır.
// İki parça: "screen" (ortada ekran + sağda yazı ve indirilenler) ve "panel" (kenar çubuğunda gezinme, adres, sekmeler).
// Ekran önce doğrudan akış kanalını (WebSocket, stream.py) dener: kareler anında gelir, dokunuşlar anında gider.
// Kanal yoksa kareler ve olaylar Streamlit üzerinden gider (daha yavaş ama her zaman çalışır).

const STATE = new WeakMap();
const KEYS = ['Enter', 'Backspace', 'Tab', 'Escape', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', 'Delete', 'Home', 'End', 'PageUp', 'PageDown'];
const esc = (x) => String(x ?? '').replace(/[&<>"']/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]));

function channel(S) {
  S.queue = []; S.seq = 0; S.born = Date.now().toString(36);
  S.send = (event) => {
    S.queue.push(event);
    S.setStateValue('input', { seq: `${S.born}-${++S.seq}`, events: S.queue.splice(0) });
    if (S.onSend) S.onSend();
  };
  S.pending = () => S.seq && S.data.applied !== `${S.born}-${S.seq}`;
}

// Düğmeler: data-t = olay türü, data-v = değer (sekme sırası, dosya adı), data-k = tuş.
function buttons(box, S) {
  box.addEventListener('click', (e) => {
    const b = e.target.closest('button[data-t], button[data-k]'); if (!b) return;
    e.preventDefault();
    if (b.dataset.k) S.send({ t: 'key', v: b.dataset.k });
    else S.send(b.dataset.v !== undefined ? { t: b.dataset.t, v: b.dataset.v } : { t: b.dataset.t });
  });
}

const FAST = new Set(['click', 'wheel', 'type', 'key']);

function downloadRows(list) {
  const rows = (list || []).map((x) => {
    const icon = x.state === 'bitti' ? '✅' : x.state === 'hata' ? '⚠️' : '⏳';
    const info = x.state === 'bitti' ? `${x.mb} MB · bilgisayarda` : x.state === 'hata' ? `inmedi: ${esc(x.error || '')}` : `iniyor… ${x.seconds} sn`;
    const use = x.state !== 'bitti' ? ''
      : x.video ? `<button type="button" class="primary" data-t="use_download" data-v="${esc(x.name)}" title="Video Stüdyosu'nda seçili açılır">🎬 Videoda kullan</button>`
      : x.text ? `<button type="button" class="primary" data-t="use_text" data-v="${esc(x.name)}" title="Haber Stüdyosu'nda ham haber olarak açılır">📰 Habere aktar</button>` : '';
    return `<div class="dl ${x.state}"><div class="name">${icon} <b title="${esc(x.name)}">${esc(x.name)}</b></div><div class="info">${info}</div>${use}</div>`;
  }).join('');
  return rows || '<div class="empty">Henüz indirme yok. İndirilenler bilgisayarın <b>İndirilenler</b> klasörüne iner.</div>';
}

function setupScreen(root, S) {
  const $ = (s) => root.querySelector(s);
  const img = $('.screen'), wrap = $('.screen-wrap'), text = $('.text'), wait = $('.wait');
  channel(S);
  S.wheel = null;
  S.onSend = () => wait.classList.add('on');
  // Doğrudan akış: gelen her kare gösterilince "ack" (en fazla 2 kare yolda; yavaş internette gecikme birikmez).
  const show = (url, done) => {
    const next = new Image();
    next.onload = () => {
      img.src = url; done();
      if (!press) { S.shift = 0; img.style.transform = ''; }
    };
    next.onerror = done;
    next.src = url;
  };
  const connect = () => {
    const info = S.data.stream;
    if (!info || S.ws || S.gone) return;
    let ws;
    try { ws = new WebSocket(`${location.protocol === 'https:' ? 'wss' : 'ws'}://${location.host}${info.path}?t=${encodeURIComponent(info.token)}`); }
    catch (e) { return; }
    ws.binaryType = 'blob'; S.ws = ws;
    ws.onmessage = (m) => {
      if (!img.isConnected) { S.gone = true; ws.close(); return; }  // sayfadan çıkıldı: akış dursun
      S.live = true;
      const url = URL.createObjectURL(m.data);
      show(url, () => { if (S.blob) URL.revokeObjectURL(S.blob); S.blob = url; if (ws.readyState === 1) ws.send('{"t":"ack"}'); });
    };
    ws.onclose = () => { S.ws = null; S.live = false; if (!S.gone) setTimeout(connect, 3000); };
  };
  S.connect = connect;
  const streamlitSend = S.send;
  S.send = (event) => {  // dokunuş/kaydırma/yazı akıştan; giriş kaydı, indirme gibi uygulama işleri Streamlit'ten
    if (S.live && S.ws && S.ws.readyState === 1 && FAST.has(event.t)) S.ws.send(JSON.stringify({ events: [event] }));
    else streamlitSend(event);
  };
  const toView = (e) => {
    const r = img.getBoundingClientRect(), [w, h] = S.data.viewport;
    return [Math.round((e.clientX - r.left) * w / r.width), Math.round((e.clientY - r.top) * h / r.height)];
  };
  const ripple = (e) => {
    const r = wrap.getBoundingClientRect(), dot = document.createElement('span');
    dot.className = 'ripple'; dot.style.left = `${e.clientX - r.left}px`; dot.style.top = `${e.clientY - r.top}px`;
    wrap.appendChild(dot); setTimeout(() => dot.remove(), 500);
  };
  // Kaydırma: parmak hareketi birikir; akış kanalında ~30 kez/sn, Streamlit yolunda ~10 kez/sn gönderilir. Görüntü o sırada parmakla birlikte kayar
  // (yeni kare gelene kadar): uzaktaki tarayıcıyı beklemeden anında tepki hissi.
  const nudge = (px) => { S.shift = Math.max(-240, Math.min(240, (S.shift || 0) + px)); img.style.transform = `translateY(${S.shift}px)`; };
  // Sayfanın başı/sonu gibi kaymayan yerde yeni kare gelmez: kaydırılan görüntü (üstte/altta boşluk) geri yerine oturur.
  const settle = () => {
    clearTimeout(S.settle);
    S.settle = setTimeout(() => {
      if (press || !S.shift) return;
      S.shift = 0; img.style.transition = 'transform .15s ease-out'; img.style.transform = '';
      setTimeout(() => { img.style.transition = ''; }, 200);
    }, 800);
  };
  const addWheel = (point, delta) => {
    if (!S.wheel) {
      S.wheel = { point, delta: 0 };
      setTimeout(() => { const w = S.wheel; S.wheel = null; if (w && Math.abs(w.delta) >= 1) S.send({ t: 'wheel', v: [...w.point, Math.round(w.delta)] }); }, S.live ? 30 : 90);
    }
    S.wheel.delta += delta;
  };

  let press = null;
  img.addEventListener('pointerdown', (e) => {
    img.focus({ preventScroll: true });
    press = { x: e.clientX, y: e.clientY, lastY: e.clientY, moved: false, point: toView(e) };
    img.setPointerCapture(e.pointerId);
  });
  img.addEventListener('pointermove', (e) => {
    if (!press) return;
    if (!press.moved && Math.hypot(e.clientX - press.x, e.clientY - press.y) > 8) press.moved = true;
    if (press.moved) {
      const scale = S.data.viewport[1] / img.getBoundingClientRect().height;
      addWheel(press.point, (press.lastY - e.clientY) * scale);
      nudge(e.clientY - press.lastY);
      press.lastY = e.clientY;
    }
  });
  const release = (e) => {
    if (press && !press.moved && e.type === 'pointerup') { ripple(e); S.send({ t: 'click', v: press.point }); }
    press = null;
    settle();
  };
  img.addEventListener('pointerup', release);
  img.addEventListener('pointercancel', release);
  img.addEventListener('wheel', (e) => {
    e.preventDefault(); addWheel(toView(e), e.deltaY);
    nudge(-e.deltaY * img.getBoundingClientRect().height / S.data.viewport[1]);
    settle();
  }, { passive: false });
  img.addEventListener('contextmenu', (e) => e.preventDefault());
  // Fiziksel klavye: ekran seçiliyken yazılanlar doğrudan tarayıcıya gider.
  img.addEventListener('keydown', (e) => {
    if (e.ctrlKey || e.metaKey || e.altKey) return;
    if (e.key.length === 1) { e.preventDefault(); S.send({ t: 'type', v: e.key }); }
    else if (KEYS.includes(e.key)) { e.preventDefault(); S.send({ t: 'key', v: e.key }); }
  });
  $('.typebar').addEventListener('submit', (e) => { e.preventDefault(); if (text.value) { S.send({ t: 'type', v: text.value }); text.value = ''; } });
  buttons($('.side'), S);
  buttons($('.login'), S);
  const downloads = $('.downloads');

  S.apply = () => {
    const d = S.data;
    wrap.style.setProperty('--ar', `${d.viewport[0]} / ${d.viewport[1]}`);
    if (!S.ws) S.connect();
    if (!S.live && d.img && d.img !== S.shown) {  // akış yoksa: Streamlit'in yolladığı kare (önce yüklenir, titreme olmasın)
      const url = d.img;
      show(url, () => { S.shown = url; });
    }
    const rows = downloadRows(d.downloads);
    if (downloads.dataset.html !== rows) { downloads.innerHTML = rows; downloads.dataset.html = rows; }
    $('.pw').style.display = d.password ? '' : 'none';
    const login = d.login || {}, box = $('.login');
    const loginHtml = login.offer
      ? `<span>🔑 <b>${esc(login.offer.site)}</b> girişi kaydedilsin mi? (${esc(login.offer.username) || 'kullanıcı adı yok'}) Sonraki girişlerde kutular kendiliğinden dolar.</span>
         <button type="button" class="primary" data-t="save_login">Kaydet</button><button type="button" data-t="dismiss_login">Hayır</button>`
      : login.filled ? `<span>🔑 Giriş bilgileri dolduruldu. Sayfadaki <b>Giriş</b> düğmesine dokun.</span>` : '';
    if (box.dataset.html !== loginHtml) { box.innerHTML = loginHtml; box.dataset.html = loginHtml; }
    box.classList.toggle('filled', !login.offer && !!login.filled);
    if (!S.pending()) wait.classList.remove('on');
  };
}

function setupPanel(root, S) {
  const $ = (s) => root.querySelector(s);
  const url = $('.url');
  channel(S);
  buttons($('.nav'), S);
  buttons($('.tabs'), S);
  $('.go').addEventListener('submit', (e) => { e.preventDefault(); if (url.value.trim()) { S.send({ t: 'goto', v: url.value.trim() }); url.blur(); } });
  const html = (box, markup) => { if (box.dataset.html !== markup) { box.innerHTML = markup; box.dataset.html = markup; } };

  S.apply = () => {
    const d = S.data;
    if (document.activeElement !== url && S.root.getRootNode().activeElement !== url) url.value = d.url || '';
    const tabs = d.tabs || [];
    html($('.tabs'), tabs.map((tab, i) => `<div class="tab${tab.active ? ' on' : ''}">
        <button type="button" class="pick" data-t="tab" data-v="${i}" title="${esc(tab.url)}">${esc(tab.title || tab.url || 'Yeni sekme')}</button>
        ${tabs.length > 1 ? `<button type="button" class="x" data-t="close_tab" data-v="${i}" title="Sekmeyi kapat">✕</button>` : ''}</div>`).join(''));
  };
}

export default function (component) {
  const { data, parentElement, setStateValue } = component;
  let S = STATE.get(parentElement);
  if (!S) {
    S = { root: parentElement, data, setStateValue };
    STATE.set(parentElement, S);
    (parentElement.querySelector('.screen') ? setupScreen : setupPanel)(parentElement, S);
  }
  S.data = data; S.setStateValue = setStateValue;
  S.apply();
}
